core/base/batch_utilities.hpp

```diff
@@ -121,14 +126,17 @@ auto unbatch(const InputType* batch_object)
 }
 
 
-template <typename ValueType, typename IndexType, typename OutputType>
+template <typename ValueType, typename IndexType, typename OutputType,
+          typename... TArgs>
 std::unique_ptr<OutputType> read(
     std::shared_ptr<const Executor> exec,
-    const std::vector<gko::matrix_data<ValueType, IndexType>>& data)
+    const std::vector<gko::matrix_data<ValueType, IndexType>>& data,
+    TArgs&&... create_args)
 {
     auto num_batch_items = data.size();
     auto tmp =
-        OutputType::create(exec, batch_dim<2>(num_batch_items, data[0].size));
+        OutputType::create(exec, batch_dim<2>(num_batch_items, data[0].size),
```

Comment:
it requires the knowledge before reading the matrix.
creating batchELL without enough num_stored_elems_per_row will lead the error.
The column index keeps changed and it will use the last one as the batch uniform sparsity
even if the last one contain the correct column index, the previous matrix may still be wrong.

Reply:
1. If the `num_stored_elems_per_row` is not passed in, then it is a dense matrix, so that should always work.
2. If `num_stored_elems_per_row` passed through `create_args` is < than the `num_stored_elems_per_row` computed from the `data[b]`, then the `read` tries to `resize` the data and fails, which IMO is acceptable.
3. If `num_stored_elems_per_row` passed through `create_args` is > than the `num_stored_elems_per_row` computed from `data[b]`, it is filled with zeros, which should always be valid.

Reply:
1. no, I think the type is based on given template OutputType?
2. 3. Yes, but there's no clue about the `num_stored_elems_per_row` should be passed
no same sparsity is still the issue

Reply:
I think you are right with the case not having same sparsity. I think I will just throw when there is non-uniform sparsity now. We can later add a function to compute the largest sparsity, but I dont think this will be a major use case. 

I think it is acceptable that the user needs to specify the largest `num_stored_elems_per_row`. If it is smaller than possible, the function throws anyway, so I think that should be okay. This will be the case for only ELL anyway, which is not a general storage format.

Reply:
And for 1. I mean the type is still OutputType, but the `num_stored_elems_per_row` is set to num_cols, which is equivalent to ELL storing a dense matrix. If the data is sparse, then we throw in the `gko::matrix::Ell::read` function.